File: leetcode/src/main/java/com/giwankim/leetcode/MaximumUnitsOnATruck.java
package com.giwankim.leetcode;

import java.util.Arrays;

public class MaximumUnitsOnATruck {
  public int maximumUnits(int[][] boxTypes, int truckSize) {
    int result = 0;
    //    Arrays.sort(boxTypes, Comparator.<int[]>comparingInt(a -> a[1]).reversed());
    Arrays.sort(boxTypes, (a, b) -> b[1] - a[1]);
    for (int[] boxType : boxTypes) {
      int take = Math.min(boxType[0], truckSize);
      result += take * boxType[1];
      truckSize -= take;
    }
    return result;
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/MaximumUnitsOnATruck.kt
package com.giwankim.leetcode

class MaximumUnitsOnATruck {
    fun maximumUnits(
        boxTypes: Array<IntArray>,
        truckSize: Int,
    ): Int {
        var result = 0
        var size = truckSize
        boxTypes.sortByDescending { it[1] }
        for (boxType in boxTypes) {
            if (boxType[0] <= size) {
                result += boxType[0] * boxType[1]
                size -= boxType[0]
            } else {
                result += size * boxType[1]
                break
            }
        }
        return result
    }
}
